index.d.ts

```diff
@@ -0,0 +1,125 @@
+    | Action<S, P, D>
+
+  // A view builds a virtual DOM node representation of the application state.
+  type View = <S>(state: State<S>) => VDOM
+
+  // Application state is accessible in every view, action, and subscription.
+  type State<S> = S
+
+  // A subscription is a set of recurring effects.
+  type Subscription = <S>(state: State<S>) => Subscriber[]
+
+  // A subscriber reacts to subscription updates.
+  type Subscriber = boolean | void | Effect | Unsubscribe
+
+  // A subscriber ideally provides a function that cancels itself properly.
+  type Unsubscribe = () => void
+
+  // Middleware allows for custom processing during dispatching.
+  type Middleware = (dispatch: Dispatch) => Dispatch
+
+  // ---------------------------------------------------------------------------
+
+  // A dispatched action handles an event in the context of the current state.
+  type Dispatch = <S, P, D>(action: Action<S, P, D>, props?: Payload<P>) => void
+
+  // An action transforms existing state while possibly invoking effects and it
+  // can be wrapped by another action.
+  type Action<S, P, D>
+    = [Action<S, P, D>, Payload<P>]
+    | ((state: State<S>, props?: Payload<P>)
+      => State<S>
+      | [State<S>, ...EffectDescriptor<D>[]]
+      | Action<S, P, D>)
+
+  // A payload is data external to state that is given to a dispatched action.
+  type Payload<P> = P
+
+  // An effect descriptor describes how an effect should be invoked.
+  // A function that creates this is called an effect constructor.
+  type EffectDescriptor<D> = [Effect, EffectData<D>]
+
+  // An effect is where side effects and any additional dispatching occur.
+  type Effect = <D>(dispatch: Dispatch, props?: EffectData<D>) => void
+
+  // An effect is generally given additional data.
+  type EffectData<D> = D
+
+  // ---------------------------------------------------------------------------
+
+  // A virtual DOM node represents an actual DOM element.
+  type VDOM = {
+    readonly type: string;
+    readonly props: PropList;
+    readonly children: VNode[];
+    node: MaybeNode;
+    readonly tag?: Tag;
+    readonly key: Key;
+    memo?: PropList;
+  }
+
+  // Virtual DOM properties will often correspond to HTML attributes.
+  type Prop = bigint | boolean | null | number | string | symbol | undefined | Function | ClassProp | StyleProp
+  type PropList = Readonly<ElementCreationOptions & {
+    [k: string]: Prop;
+    class?: ClassProp;
+    key?: Key;
+    style?: StyleProp;
+  }>
+
+  // A key can uniquely associate a virtual DOM node with a certain DOM element.
+  type Key = null | string | undefined
+
+  // The `class` property represents an HTML class attribute string.
+  type ClassProp = string | Record<string, boolean> | ClassProp[]
+
+  // The `style` property represents inline CSS.
+  type StyleProp = Record<string, null | number | string>
+
+  // A virtual node is a convenience layer over a virtual DOM node.
+  type VNode = null | undefined | VDOM
```

Comment:
@icylace `VNode` also accepts `boolean` which are filtered out down the road. Here's an example:

```
view: () => h('div', {}, [false])
```
and
```
view: () => h('div', {}, false)
```

Reply:
What happens with `true` ?

Reply:
Nothing happens. Either `true` or `false` the node (in this case `div`) is not created.